test_conformance/api/test_api_consistency.cpp

```diff
@@ -473,14 +437,688 @@ int test_consistency_pipes(cl_device_id deviceID, cl_context context,
+            clCreateProgramWithIL(context, &bogus, sizeof(bogus), &error);
+        test_failure_error(
+            error, CL_INVALID_OPERATION,
+            "Device does not support IL Programs but clCreateProgramWithIL did "
+            "not return CL_INVALID_OPERATION");
+
+        // clSetProgramSpecializationConstant
+        // Returns CL_INVALID_OPERATION if no devices associated with program
+        // support Intermediate Language Programs.
+
+        cl_uint specConst = 42;
+        error = clSetProgramSpecializationConstant(
+            program, 0, sizeof(specConst), &specConst);
+        test_failure_error(error, CL_INVALID_OPERATION,
+                           "Device does not support IL Programs but "
+                           "clSetProgramSpecializationConstant did not return "
+                           "CL_INVALID_OPERATION");
+    }
+
+    return TEST_PASS;
+}
+
+int test_consistency_subgroups(cl_device_id deviceID, cl_context context,
+                               cl_command_queue queue, int num_elements)
+{
+    // clGetDeviceInfo, passing CL_DEVICE_MAX_NUM_SUB_GROUPS
+    // May return 0, indicating that device does not support Subgroups.
+    cl_int error;
+
+    clProgramWrapper program;
+    clKernelWrapper kernel;
+
+    cl_uint maxNumSubGroups = 0;
+    error = clGetDeviceInfo(deviceID, CL_DEVICE_MAX_NUM_SUB_GROUPS,
+                            sizeof(maxNumSubGroups), &maxNumSubGroups, NULL);
+    test_error(error, "Unable to query CL_DEVICE_MAX_NUM_SUB_GROUPS");
+
+    if (maxNumSubGroups == 0)
+    {
+        // Test setup:
+
+        error = create_single_kernel_helper(context, &program, &kernel, 1,
+                                            &test_kernel, "test");
+        test_error(error, "Unable to create test kernel");
+
+        // clGetDeviceInfo, passing
+        // CL_DEVICE_SUB_GROUP_INDEPENDENT_FORWARD_PROGRESS
+        // Returns CL_FALSE if device does not support Subgroups.
+
+        cl_bool ifp = CL_FALSE;
+        error = clGetDeviceInfo(
+            deviceID, CL_DEVICE_SUB_GROUP_INDEPENDENT_FORWARD_PROGRESS,
+            sizeof(ifp), &ifp, NULL);
+        test_error(
+            error,
+            "Unable to query CL_DEVICE_SUB_GROUP_INDEPENDENT_FORWARD_PROGRESS");
+        test_assert_error(ifp == CL_FALSE,
+                          "Device does not support Subgroups but "
+                          "CL_DEVICE_SUB_GROUP_INDEPENDENT_FORWARD_PROGRESS "
+                          "did not return CL_FALSE");
+
+        // clGetDeviceInfo, passing CL_DEVICE_EXTENSIONS
+        // Will not describe support for the cl_khr_subgroups extension if
+        // device does not support Subgroups.
+
+        bool supports_cl_khr_subgroups =
+            is_extension_available(deviceID, "cl_khr_subgroups");
+        test_assert_error(supports_cl_khr_subgroups == false,
+                          "Device does not support Subgroups but does "
+                          "support cl_khr_subgroups");
+
```

Comment:
What do you think about adding condition to check if supports_cl_khr_subgroups == cl_true then ifp == cl_true ?

Reply:
I thought this check would have existed already, but it doesn't look like it does.

I don't mind adding it but would you be OK adding it in a separate PR?  I'm going to need to make a minor modification to this test for #933 anyhow.

Reply:
We have such verification in subgroups test https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/subgroups/test_ifp.cpp#L356 . 
Then maybe there is no need to add such verification here again.